hudi-flink-writer/src/main/java/org/apache/hudi/operator/InstantGenerateOperator.java

```diff
@@ -0,0 +1,219 @@
+    // check whether the last instant is completed, if not, wait 10s and then throws an exception
+    if (!StringUtils.isNullOrEmpty(latestInstant)) {
+      doChecker();
+      // last instant completed, set it empty
+      latestInstant = "";
+    }
+
+    // no data no new instant
+    if (!bufferedRecords.isEmpty()) {
+      latestInstant = startNewInstant(checkpointId);
+    }
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    // instantState
+    ListStateDescriptor<String> latestInstantStateDescriptor = new ListStateDescriptor<String>("latestInstant", String.class);
+    latestInstantState = context.getOperatorStateStore().getListState(latestInstantStateDescriptor);
+
+    // recordState
+    ListStateDescriptor<StreamRecord> recordsStateDescriptor = new ListStateDescriptor<StreamRecord>("recordsState", StreamRecord.class);
+    recordsState = context.getOperatorStateStore().getListState(recordsStateDescriptor);
+
+    if (context.isRestored()) {
+      Iterator<String> latestInstantIterator = latestInstantState.get().iterator();
+      latestInstantIterator.forEachRemaining(x -> latestInstant = x);
+      LOG.info("InstantGenerateOperator initializeState get latestInstant [{}]", latestInstant);
+
+      Iterator<StreamRecord> recordIterator = recordsState.get().iterator();
+      bufferedRecords.clear();
+      recordIterator.forEachRemaining(x -> bufferedRecords.add(x));
+    }
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext functionSnapshotContext) throws Exception {
+    if (latestInstantList.isEmpty()) {
+      latestInstantList.add(latestInstant);
+    } else {
+      latestInstantList.set(0, latestInstant);
+    }
+    latestInstantState.update(latestInstantList);
+    LOG.info("Update latest instant [{}]", latestInstant);
+
+    recordsState.update(bufferedRecords);
+    LOG.info("Update records state size = [{}]", bufferedRecords.size());
+    bufferedRecords.clear();
+  }
+
+  /**
+   * Create a new instant.
+   *
+   * @param checkpointId
+   */
+  private String startNewInstant(long checkpointId) {
+    String newTime = writeClient.startCommit();
+    LOG.info("create instant [{}], at checkpoint [{}]", newTime, checkpointId);
+    return newTime;
+  }
+
+  /**
+   * Check the status of last instant.
+   */
+  private void doChecker() throws InterruptedException {
+    // query the requested and inflight commit/deltacommit instants
+    String commitType = cfg.tableType.equals(HoodieTableType.COPY_ON_WRITE.name()) ? HoodieTimeline.COMMIT_ACTION : HoodieTimeline.DELTA_COMMIT_ACTION;
+    LOG.info("Query latest instant [{}]", latestInstant);
+    List<String> rollbackPendingCommits = writeClient.getInflightsAndRequestedInstants(commitType);
+    int tryTimes = 0;
+    while (tryTimes < commitTimeout) {
+      tryTimes++;
+      StringBuffer sb = new StringBuffer();
+      if (rollbackPendingCommits.contains(latestInstant)) {
+        //清空 sb
```

Comment:
Chinese? Why this is a clear action to string buffer?

Reply:
> Chinese? Why this is a clear action to string buffer?

my bad